pylabnet/scripts/pulsemaster/pulsemaster.py

```diff
@@ -0,0 +1,1457 @@
+        input_widget_name = f"{slugify(field_dict['label'])}_{str(field_input_str)}"
+
+        if pulse_mod:
+            input_widget_name = f"{input_widget_name}_pulse_mod"
+
+        return input_widget_name
+
+
+    def _get_pulse_fieldtype(self, field_dict):
+        """Determie input field type."""
+        # Build field.
+        field_type = field_dict['input_type']
+        if field_type == 'QLineEdit':
+            field_input = QLineEdit()
+        elif field_type == 'QComboBox':
+            field_input = QComboBox()
+        return field_input
+
+    def build_pulse_input_fields(self):
+        """Change input fields if pulse selector dropdown has been changed."""
+
+        # Remove old entries.
+        self._remove_variable_pulse_fields()
+
+        # Load pulsetype settings
+        pulsetype_dict = self._current_pulsetype_dict()
+
+        for field in pulsetype_dict['fields']:
+
+            # Add label.
+            field_label = QLabel(field['label'])
+
+            # Get field type
+            field_input = self._get_pulse_fieldtype(field)
+
+            # Add choices to combobox.
+            if type(field_input) is QComboBox:
+                    field_input.addItems(field['combo_choice'])
+
+            # Auto create name of widget:
+            input_widget_name = self._get_form_field_widget_name(field)
+            field_input.setObjectName(input_widget_name)
+
+            self.pulse_selector_form_layout_variable.addRow(field_label, field_input)
+
+        # Apply CSS stylesheet
+        self.gui.apply_stylesheet()
+
+
+    def set_pulsetype_combobox(self, combobox):
+        for pulsetype in self.config_dict['pulse_types']:
+            combobox.addItem(pulsetype['name'])
+
+
+    def set_dio_channel_completer(self):
+        """Reset the autocomplete for the channel selection."""
+        completer = QCompleter(self.DIO_assignment_dict.keys())
+        self.pulse_selector_channelselection.setCompleter(completer)
+
+    def load_dio_assignment_from_dict(self):
+        """Read in DIO assignment dictionary and store as member variable."""
+        # Load DIO assignment.
+        self.DIO_assignment_dict = load_config(
+                config_filename=self.config_dict['DIO_dict'],
+                logger=self.log
+        )
+
+    def populate_dio_table_from_dict(self):
+        '''Populate DIO assignment table from DIO assignment dict.'''
+
+        # Update DIO assignments from dict
+        self.load_dio_assignment_from_dict()
+
+        self.model = DictionaryTableModel(self.DIO_assignment_dict, header=["Channel Name", "DIO Bit"])
+        self.widgets['DIO_table'].setModel(self.model)
+
+        # Update completer.
+        self.set_dio_channel_completer()
+
+        self.log.info('DIO settings successfully loaded.')
+
+    def run(self):
+        """ Runs an iteration of checks for updates and implements
+        """
+
+
+        time.sleep(0.01)
+        self.gui.force_update()
```

Comment:
i found out that (by reading yan qi and pieter's code mainly) that you can just call sel.gui.app._exec() instead, and avoid a while loop that continually forces updates. supposedly this may put less strain on the CPU. The `force_update()` which calls `app.processEvents()` is mainly for processes triggered by the GUI in the main GUI thread which may be time consuming. In this case, you want to `app.processEvents()` during the execution of this time-consuming process (or just move the process to another thread)

Reply:
Ok, unsure how to actually change the code without breaking it, let's chat about is later.

Reply:
I think you can just get rid of the `run()` function altogether, as well as the `while` loop in `launch()` and in the last line of the launcher run `pulsemaster.gui.app.exec_()`. Below is how it is done in the `data_taker` module:
```python
def launch(**kwargs):

    logger, loghost, logport, clients, guis, params = unpack_launcher(**kwargs)

    # Instantiate Monitor script
    control = DataTaker(
        logger=logger,
        clients=clients,
        config=kwargs['config'],
    )
    control.gui.set_network_info(port=kwargs['server_port'])

    # Run continuously
    control.gui.app.exec_()
```

Reply:
Yep, seems to work!